ecosystem/typescript/sdk_v2/src/bcs/serializer.ts

```diff
@@ -0,0 +1,308 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+/* eslint-disable no-bitwise */
+import {
+  MAX_U128_BIG_INT,
+  MAX_U16_NUMBER,
+  MAX_U32_NUMBER,
+  MAX_U64_BIG_INT,
+  MAX_U8_NUMBER,
+  MAX_U256_BIG_INT,
+} from "./consts";
+import { AnyNumber, Uint16, Uint32, Uint8 } from "./types";
+
+export interface Serializable {
+  serialize(serializer: Serializer): void;
+}
+
+export class Serializer {
+  private buffer: ArrayBuffer;
+
+  private offset: number;
+
+  constructor() {
+    this.buffer = new ArrayBuffer(64);
+    this.offset = 0;
+  }
+
+  private ensureBufferWillHandleSize(bytes: number) {
+    while (this.buffer.byteLength < this.offset + bytes) {
+      const newBuffer = new ArrayBuffer(this.buffer.byteLength * 2);
+      new Uint8Array(newBuffer).set(new Uint8Array(this.buffer));
+      this.buffer = newBuffer;
+    }
+  }
+
+  protected appendToBuffer(values: Uint8Array) {
+    this.ensureBufferWillHandleSize(values.length);
+    new Uint8Array(this.buffer, this.offset).set(values);
+    this.offset += values.length;
+  }
+
+  private serializeWithFunction(
+    fn: (byteOffset: number, value: number, littleEndian?: boolean) => void,
+    bytesLength: number,
+    value: number,
+  ) {
+    this.ensureBufferWillHandleSize(bytesLength);
+    const dv = new DataView(this.buffer, this.offset);
+    fn.apply(dv, [0, value, true]);
+    this.offset += bytesLength;
+  }
+
+  /**
+   * Serializes a string. UTF8 string is supported.
+   *
+   * The number of bytes in the string content is serialized first, as a uleb128-encoded u32 integer.
+   * Then the string content is serialized as UTF8 encoded bytes.
+   *
+   * BCS layout for "string": string_length | string_content
+   * where string_length is a u32 integer encoded as a uleb128 integer, equal to the number of bytes in string_content.
+   *
+   * @example
+   * ```ts
+   * const serializer = new Serializer();
+   * serializer.serializeStr("1234abcd");
+   * assert(serializer.getBytes() === new Uint8Array([8, 49, 50, 51, 52, 97, 98, 99, 100]));
+   * ```
+   */
+  serializeStr(value: string): this {
+    const textEncoder = new TextEncoder();
+    this.serializeByteVector(textEncoder.encode(value));
+    return this;
+  }
+
+  /**
+   * Serializes an array of bytes.
+   *
+   * BCS layout for "bytes": bytes_length | bytes
+   * where bytes_length is a u32 integer encoded as a uleb128 integer, equal to the length of the bytes array.
+   */
+  serializeByteVector(value: Uint8Array): this {
```

Comment:
I noticed that the name in v1 is `serializeBytes`, shall we keep it the same? I think Bytes and Uint8Array feel much the same.

Reply:
yeah I might go back to it, the only reason I changed it is because in the old SDK `serializeBytes` is confusing, because there's also `serializeFixedBytes` and `bcsToBytes`.

They all do 3 different things.

Where as `serializeByteVector` is very clear that you're serializing a representation of `vector<u8>`